Move mandatory inlining cleanup to mandatory combine

zoecarver opened the pull request:
<!-- What's in this pull request? -->
This patch makes `MandatoryInlining` a `SILModuleTransform` instead of a `SILFunctionTransform`. It also moves the function removal cleanup to `MandatoryCombine` so that it will be run on every function. This is better because then we don't have to iterate over every function in a module twice in `MandatoryInlining`. And now, both can remain a `SILFunctionTransform`. It also has the benefit of allowing `MandatoryCombine` to exit early sometimes. 

The successor to #28394.


<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

cc @gottesmm

zoecarver committed:
Make MandatoryInlining a SILFunctionTransform

zoecarver committed:
Move mandatory inlining cleanup to mandatory combine

gottesmm commented:
This is the wrong way to do this. What we want to do is move stuff into the mandatory combiner since it runs after diagnostics. That way we can make it easy for SourceKit to not run this sort of code.

zoecarver committed:
Re-write recursive inliner to run on single function

zoecarver committed:
Fix some tests

zoecarver committed:
Move closure cleanup into loop

zoecarver committed:
Cleanup & comments

zoecarver commented:
I re-wrote this patch fixing most of the issues that you brought up. I removed changes to `MandatoryCombine`. I'll either add those to this patch later or add them in another patch. 

The basic flow of this pass is now:
* Collect apply instructions.
* Try to devitalize applys.
* Iterate over every apply and try to inline it. Otherwise, continue.
* Add the apply's callee to the pass manager worklist so it will also try to be inlined (instead of trying to inline it recursively).
* If any functions were inlined, add them to the worklist as well so the now-dead functions can be removed.

The recursive inlining is now removed. All mandatory inlining tests pass (assuming I ran the right tests). 

There are still a few things I want to clean up (see TODOs). I'll probably fix those in a different patch, though.

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -764,137 +764,89 @@ static SILInstruction *tryDevirtualizeApplyHelper(FullApplySite InnerAI,
+
+  // First collect all the applies
+  SmallVector<FullApplySite, 12> applyInsts;
+  while (SILBasicBlock *block = domOrder.getNext()) {
```
Why do you need to use the dominator info here?

zoecarver replied:
We don't. I'll remove it.

zoecarver replied:
Hmm. To be honest, I'm not quite sure why reading the blocks in dominance order fixes this, but, when inlining an auto closure with `-enable-ownership-stripping-after-serialization`, the verifier fails because `convert_escape_to_noescape [not_guaranteed] not allowed after mandatory passes`. Using dominance order fixes that.

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -764,137 +764,89 @@ static SILInstruction *tryDevirtualizeApplyHelper(FullApplySite InnerAI,
-    return false;
+static size_t collectApplies(SILFunction *F, ClassHierarchyAnalysis *CHA,
+                             DominanceAnalysis *DA,
+                             SmallVector<FullApplySite, 8> &collection) {
```
SmallVector -> SmallVectorImpl.

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -764,137 +764,89 @@ static SILInstruction *tryDevirtualizeApplyHelper(FullApplySite InnerAI,
+    collection.push_back(apply);
+  }
+
+  return collection.size();
```
Why not just return a bool?

zoecarver replied:
I was using `.size` as `!.empty`. I'll update it.

gottesmm commented on lib/SILOptimizer/PassManager/PassPipeline.cpp:
```diff
@@ -606,6 +607,7 @@ SILPassPipelinePlan::getSILOptPreparePassPipeline(const SILOptions &Options) {
   }
 
   P.startPipeline("SILOpt Prepare Passes");
+  P.addMandatoryInlining();
```
You shouldn't need another run here.

zoecarver replied:
I thought this was needed but you're right (it isn't).

gottesmm commented on lib/SILOptimizer/PassManager/PassPipeline.cpp:
```diff
@@ -117,6 +117,7 @@ static void addMandatoryOptPipeline(SILPassPipelinePlan &P) {
   if (!Options.StripOwnershipAfterSerialization)
     P.addOwnershipModelEliminator();
   P.addMandatoryInlining();
+  P.addMandatoryCombine();
```
Please remove this.

gottesmm commented on lib/SILOptimizer/PassManager/PassPipeline.cpp:
```diff
@@ -665,6 +667,7 @@ SILPassPipelinePlan::getOnonePassPipeline(const SILOptions &Options) {
   SILPassPipelinePlan P(Options);
 
   P.startPipeline("Mandatory Combines");
+  P.addMandatoryInlining();
```
This should not be necessary. Please remove this.

zoecarver replied:
[`testOriginalProtocolImpl` doesn't inline `cFuncOriginal` in `mandatory_conditional_compile_out_using_optionals.swift`](https://github.com/apple/swift/blob/master/test/SILOptimizer/mandatory_conditional_compile_out_using_optionals.swift#L73) (among others) when this pass isn't run for `Onone`.

gottesmm reviewed (changes requested):
Some initial comments. I want to read this code in more detail before it lands.

zoecarver committed:
Fix based on review

gottesmm commented:
So I was thinking about this a little bit. We need to do this differently/split this up. Here is my suggestion:

1. Why don't we create a separate PR that just moves cleanup code (and leaves everything else the same). This would mean eliminating the code in cleanup callee and moving that to the Mandatory Combiner. No passes should be moved around, it should just require having the mandatory inliner not cleaning up after itself and leaving that later for the mandatory combiner.

2. There is more complexity here around the pass manager. I can explain it in more detail offline, but I think that this may be a bit more complex than a simple cleanup. Maybe we do 1 in the short term and then circle back around to this?

zoecarver commented:
@gottesmm, that sounds good to me. 

I'll look into making a cleanup patch. I'm not 100% sure the cleanup done here is necessary, though. I think the mandatory combiner might already handle this. I'll double-check and make a patch if not. 

I would also be interested to learn more about the pass manager. Feel free to reach out to me whenever (no rush, though, I'm sure you're busy).

zoecarver commented:
Closing for now. Might reopen later.
